external/fv3fit/fv3fit/_shared/pytorch/graph.py

```diff
@@ -0,0 +1,205 @@
+        default_factory=lambda: BuildingGraph()
+    )
+
+    graph_network: GraphNetworkConfig = dataclasses.field(
+        default_factory=lambda: GraphNetworkConfig()
+    )
+    training_loop: TrainingLoopConfig = dataclasses.field(
+        default_factory=lambda: TrainingLoopConfig()
+    )
+    loss: LossConfig = LossConfig(loss_type="mse")
+
+    @property
+    def variables(self) -> Set[str]:
+        return set(self.input_variables).union(self.output_variables)
+
+
+# Some temporary transforms to make life easy for training process
+@curry
+def select_keys_mapping(
+    variable_names: Sequence[str], data: Mapping[str, tf.Tensor]
+) -> Mapping[str, tf.Tensor]:
+    return {name: data[name] for name in variable_names}
+
+
+@curry
+def normalize(
+    means: Mapping[str, float], stds: Mapping[str, float], data: Mapping[str, tf.Tensor]
+) -> Mapping[str, tf.Tensor]:
+    normed_data = {}
+    for key, tensor in data.items():
+        normed_data[key] = (tensor - means[key]) / stds[key]
+
+    return normed_data
+
+
+# TODO: Still have to handle forcing
+
+
+@register_training_function("graph", GraphHyperparameters)
+def train_graph_model(
+    hyperparameters: GraphHyperparameters,
+    train_batches: tf.data.Dataset,
+    validation_batches: Optional[tf.data.Dataset],
+):
+
+    """
+    Train a graph network.
+
+    Args:
+        train_batches: training data, as a dataset of Mapping[str, tf.Tensor]
+        validation_batches: validation data, as a dataset of Mapping[str, tf.Tensor]
+        build_model: the function which produces the pytorch model
+            from input and output samples. The models returned must take a list of
+            tensors as input and return a list of tensors as output.
+        input_variables: names of inputs for the pytorch model
+        output_variables: names of outputs for the pytorch model
+        n_epoch: number of epochs
+    """
+
+    # use transforms to get correct variables, correct dimensions
+    processed_train_batches = train_batches.map(apply_to_mapping(ensure_nd(3)))
+
+    sample = next(
+        iter(
+            processed_train_batches.unbatch().batch(
+                hyperparameters.training_loop.build_samples
+            )
+        )
+    )
+
+    # create a a transform that will properly normalize or denormalize values
+    means = apply_to_mapping(np.mean, sample)
+    stds = apply_to_mapping(np.std, sample)
+    processed_train_batches = processed_train_batches.map(normalize(means, stds))
+
+    get_Xy = curry(get_Xy_dataset)(
+        input_variables=hyperparameters.input_variables,
+        output_variables=hyperparameters.output_variables,
+        n_dims=3,
+    )
+
+    if validation_batches is not None:
+        validation_batches = validation_batches.map(normalize(means, stds))
+
+        val_Xy = get_Xy(data=validation_batches)
+    else:
+        val_Xy = None
+
+    train_Xy = get_Xy(data=processed_train_batches)
```

Comment:
Why are train batches "processed" but validation batches are not?

You could cleanly apply the same logic to both if you move the normalization logic into `get_Xy`.